File: 0874-backspace-string-compare/0874-backspace-string-compare.java
class Solution {
    private int nextValidIndex(String s, int idx) {
        int skip = 0;
        while (idx >= 0) {
            if (s.charAt(idx) == '#') {
                skip++;
                idx--;
            } else if (skip > 0) {
                skip--;
                idx--;
            } else break;
        }

        return idx;
    }

    public boolean backspaceCompare(String s, String t) {
        int i = s.length() - 1;
        int j = t.length() - 1;

        while (i >= 0 || j >= 0) {
            i = nextValidIndex(s, i);
            j = nextValidIndex(t, j);

            if (i < 0 && j < 0) return true;
            if (i < 0 || j < 0) return false;
            if (s.charAt(i) != t.charAt(j)) return false;

            i--;
            j--;
        }

        return true;
    }
}

File: 0874-backspace-string-compare/0874-backspace-string-compare.kt
class Solution {
    private fun nextValidIndex(s: String, index: Int): Int {
        var idx = index
        var skip = 0

        while (idx >= 0) {
            when {
                s[idx] == '#' -> {
                    skip++
                    idx--
                }
                skip > 0 -> {
                    skip--
                    idx--
                }
                else -> break
            }
        }

        return idx
    }

    fun backspaceCompare(s: String, t: String): Boolean {
        var i = s.length - 1
        var j = t.length - 1

        while (i >= 0 || j >= 0) {
            i = nextValidIndex(s, i)
            j = nextValidIndex(t, j)

            if (i < 0 && j < 0) return true
            if (i < 0 || j < 0) return false
            if (s[i] != t[j]) return false

            i--
            j--
        }

        return true
    }
}